src/data/equiv/derangements/basic.lean

```diff
@@ -0,0 +1,223 @@
+/-
+Copyright (c) 2021 Henry Swanson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henry Swanson
+-/
+import data.equiv.basic
+import data.equiv.option
+import group_theory.perm.option
+
+import tactic.equiv_rw
+
+/-!
+# Derangements on types
+
+In this file we define a predicate and its corresponding set:
+
+* `is_derangement f`: whether the permutation `f` has no fixed points.
+
+* `derangements α`: the set of derangements on a type `α`.
+
+Then we define
+
+* some related predicates: e.g.,
+
+  - `only_possible_fixed_point f a`: either `f` has no fixed points, or `a` is the only fixed point;
+
+  - `exactly_one_fixed_point f a`: `f` has exactly one fixed point, which is `a`;
+
+* equivalences involving various subtypes of `perm α` and `derangements α`: e.g.,
+
+  - `derangements_equiv_sigma_opfp`: an equivalence between `derangements (option α)` and the
+    sigma-type `Σ a : α, {f : perm α // only_possible_fixed_point f a}`.
+-/
+
+open equiv
+
+namespace perm
+
+section definitions
+
+/-- A permutation is a derangement if it has no fixed points. -/
+def is_derangement {α : Type*} (f : perm α) : Prop := ∀ x : α, f x ≠ x
+
+/-- The set of derangements on `α`. -/
+def derangements (α : Type*) : set (perm α) := {f : perm α | is_derangement f}
```

Comment:
`is_derangement` and `derangements` are quite literally the same thing. A term of type `set (perm α)` is the same as a function `perm α → Prop`. And you've defined them so that they have equal values. I suggest you keep only one of them, depending on whether you care about set notation or not (I think in this particular case you do).